src/analysis/inner_product_space/rayleigh.lean

```diff
@@ -0,0 +1,278 @@
+begin
+  obtain ⟨a, b, h₁, h₂⟩ := hT.linearly_dependent_of_is_local_extr_on hextr,
+  by_cases hx₀ : x₀ = 0,
+  { simp [hx₀] },
+  by_cases hb : b = 0,
+  { have : a ≠ 0 := by simpa [hb] using h₁,
+    refine absurd _ hx₀,
+    apply smul_right_injective F this,
+    simpa [hb] using h₂ },
+  let c : ℝ := - b⁻¹ * a,
+  have hc : T x₀ = c • x₀,
+  { have : b * (b⁻¹ * a) = a := by field_simp [mul_comm],
+    apply smul_right_injective F hb,
+    simp [c, ← neg_eq_of_add_eq_zero h₂, ← mul_smul, this] },
+  convert hc,
+  have : ∥x₀∥ ≠ 0 := by simp [hx₀],
+  field_simp,
+  simpa [inner_smul_left, real_inner_self_eq_norm_sq, sq] using congr_arg (λ x, ⟪x, x₀⟫_ℝ) hc,
+end
+
+end real
+
+section complete_space
+variables [complete_space E] {T : E →L[𝕜] E}
+
+lemma eq_smul_self_of_is_local_extr_on (hT : is_self_adjoint (T : E →ₗ[𝕜] E)) {x₀ : E}
+  (hextr : is_local_extr_on T.re_apply_inner_self (sphere (0:E) ∥x₀∥) x₀) :
+  T x₀ = (↑(T.re_apply_inner_self x₀ / ∥x₀∥ ^ 2) : 𝕜) • x₀ :=
+begin
+  letI := inner_product_space.is_R_or_C_to_real 𝕜 E,
+  letI : is_scalar_tower ℝ 𝕜 E := restrict_scalars.is_scalar_tower _ _ _,
+  let S : E →L[ℝ] E :=
+    @continuous_linear_map.restrict_scalars 𝕜 E E _ _ _ _ _ _ _ ℝ _ _ _ _ T,
+  have hSA : is_self_adjoint (S : E →ₗ[ℝ] E) := λ x y, by
+  { have := hT x y,
+    simp only [continuous_linear_map.coe_coe] at this,
+    simp only [real_inner_eq_re_inner, this, continuous_linear_map.coe_restrict_scalars,
+      continuous_linear_map.coe_coe, linear_map.coe_restrict_scalars_eq_coe] },
+  exact eq_smul_self_of_is_local_extr_on_real hSA hextr,
+end
+
+/-- For a self-adjoint operator `T`, a local extremum of the Rayleigh quotient of `T` on a sphere
+centred at the origin is an eigenvector of `T`. -/
+lemma has_eigenvector_of_is_local_extr_on (hT : is_self_adjoint (T : E →ₗ[𝕜] E)) {x₀ : E}
+  (hx₀ : x₀ ≠ 0) (hextr : is_local_extr_on T.re_apply_inner_self (sphere (0:E) ∥x₀∥) x₀) :
+  has_eigenvector (T : E →ₗ[𝕜] E) ↑(T.re_apply_inner_self x₀ / ∥x₀∥ ^ 2) x₀ :=
+begin
+  refine ⟨_, hx₀⟩,
+  rw module.End.mem_eigenspace_iff,
+  exact hT.eq_smul_self_of_is_local_extr_on hextr
+end
+
+/-- For a self-adjoint operator `T`, a maximum of the Rayleigh quotient of `T` on a sphere centred
+at the origin is an eigenvector of `T`, with eigenvalue the global supremum of the Rayleigh
+quotient. -/
+lemma has_eigenvector_of_is_max_on (hT : is_self_adjoint (T : E →ₗ[𝕜] E)) {x₀ : E}
+  (hx₀ : x₀ ≠ 0) (hextr : is_max_on T.re_apply_inner_self (sphere (0:E) ∥x₀∥) x₀) :
+  has_eigenvector (T : E →ₗ[𝕜] E)
+    ↑(⨆ x : {x : E // x ≠ 0}, T.re_apply_inner_self x / ∥(x:E)∥ ^ 2) x₀ :=
+begin
+  convert hT.has_eigenvector_of_is_local_extr_on hx₀ (or.inr hextr.localize),
+  have hx₀' : 0 < ∥x₀∥ := by simp [hx₀],
+  have hx₀'' : x₀ ∈ sphere (0:E) (∥x₀∥) := by simp,
+  rw T.supr_rayleigh_eq_supr_rayleigh_sphere hx₀',
+  convert is_max_on.supr_eq hx₀'' _,
+  { refl },
+  intros x hx,
+  dsimp,
+  convert div_le_div_of_le (sq_nonneg ∥x₀∥) (hextr hx) using 3,
```

Comment:
This is the first time I've seen this `convert ... using ...` trick. Very neat!

I worry it might be brittle though. Is the following worth considering:
```suggestion
  suffices : ∥x∥ = ∥x₀∥, { rw this,  exact div_le_div_of_le (sq_nonneg ∥x₀∥) (hextr hx), },
```

Reply:
Yes, sounds good.  But I made it a `have` rather than a `suffices` since the proof of `∥x∥ = ∥x₀∥` is shorter than the proof of the goal using `∥x∥ = ∥x₀∥`.